### PR TITLE
feat: implement promotion balance

### PR DESCRIPTION
**Pull request by sihamais**
### Changes proposed in this pull request
- Add `Balance` entity and controllers logic
- Add `Balance` endpoints on readme
- Add new endpoint to fetch a shop's promotions
- Add endpoint to checkout a balance (counter +1)
- Update readme
- Refactor docker config
- Refactor server config
- Write tests for new endpoints

[FID-64]

[FID-64]: https://msc-etna.atlassian.net/browse/FID-64?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,7 +28,12 @@
...
     - [Delete a promotion](#delete-a-promotion)
-
+  - [Balances endpoints](#balances-endpoints)
+    - [Get a balance](#get-a-balance)
+    - [Create a balance](#create-a-balance)
+    - [Update a balance](#update-a-balance)
+    - [Delete a balance](#delete-a-balance)
+    - [Checkout a balance](#checkout-a-balance)
 
 ## Run locally
```
[the remark this answers is not in the document]
Comment (sihamais) [reply README.md:38]: Une eternité

File: src/repositories/balance.repository.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import { Balance } from "../entities";
+
+export class BalanceRepository {
+  /**
+   * Find balance by id
+   * @param id - Id of the balance
+   * @returns A balance if found
+   */
+  static findOneById = async (id: number): Promise<Balance> => {
+    return await getDataSource()
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/repositories/balance.repository.ts]: Done

File: src/repositories/balance.repository.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        id: Number(id),
+      });
+  };
+
+  /**
+   * Save a balance on the db
+   * @param balance - The balance to save
+   */
+  static save = async (balance: Balance) => {
+    await getDataSource().getRepository(Balance).save(balance);
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/repositories/balance.repository.ts]: Done

File: src/repositories/shop.repository.ts [hunk 4]
```diff
@@ -49,4 +49,17 @@ export class ShopRepository {
   static delete = async (id: number) => {
     await getDataSource().getRepository(Shop).delete(id);
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/repositories/shop.repository.ts]: Done

File: src/repositories/shop.repository.ts [hunk 5]
```diff
@@ -49,4 +49,17 @@ export class ShopRepository {
...
   };
+
+  /**
+   * Get a shop's promotions from the db
+   * @param id - The id of shop
+   */
+  static getShopsPromotions = async (id: number): Promise<Shop> => {
+    return await getDataSource()
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/repositories/shop.repository.ts]: Done

File: src/controllers/balance.controller.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    try {
+      promotion = await PromotionRepository.findOneById(balance.promotionId);
+    } catch (error) {
+      res.status(404).send({ message: "Promotion not found" });
+      return;
+    }
+
+    if (!promotion.isActive) {
+      balance.isActive = false;
+      await BalanceRepository.save(balance);
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/controllers/balance.controller.ts]: Done

File: src/controllers/balance.controller.ts [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+      res.status(403).send({ message: "Promotion is not active" });
+      return;
+    }
+
+    if (new Date(promotion.endAt).getTime() < new Date().getTime()) {
+      promotion.isActive = false;
+      await PromotionRepository.save(promotion);
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/controllers/balance.controller.ts]: Done

File: src/controllers/balance.controller.ts [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+      res.status(403).send({ message: "Promotion has expired" });
+      return;
+    }
+
+    if (balance.counter === promotion.checkoutLimit) {
+      balance.counter = 1;
+      balance.isActive = true;
+      await BalanceRepository.save(balance);
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/controllers/balance.controller.ts:209]: Done

File: src/controllers/balance.controller.ts [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+      res.status(200).send({ message: "Balance updated" });
+      return;
+    }
+
+    if (balance.counter++ === promotion.checkoutLimit) {
+      balance.counter = promotion.checkoutLimit;
+      balance.isActive = false;
+      await BalanceRepository.save(balance);
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/controllers/balance.controller.ts]: Done

File: src/controllers/balance.controller.ts [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+      res.status(200).send({ message: "Promotion limit reached" });
+    }
+
+    balance.counter++;
+    await BalanceRepository.save(balance);
```
[the remark this answers is not in the document]
Comment (sihamais) [reply src/controllers/balance.controller.ts]: Done

**Review by AndiManLu — APPROVED** [review]
OK

**Pull request merged**